### ISSUE TITLE
QueryInterface: support for NULLS FIRST and LAST sorting in orderBy

### ISSUE THREAD

**Issue by manquer** [yiisoft/yii2#5120]
It would be great if support for explict NULL sorting in orderBy is implemented 
while not all SQL vendors implement this functionality it is defined by SQL:2003 standard extension T611, "Elementary OLAP operations" and some vendors do implement it ( Oracle and PostgreSQL both have ) and there are workaround methods for MySQL and MSSQL etc.

**Comment by gadd33** [comment]
In ActiveDataProvider-> sort -> defaultOrder this method doesn't work. Any solutions?

**Comment by arogachev** [comment]
Set the ```sort``` ```attributes```, in ```defaultOrder``` use just names of attributes, for example:

```php
defaultOrder => ['current_week_count' => SORT_ASC],
```

You should not pass db expressions in default order.

**Comment by gadd33** [comment]
I need 'DESC NULLS LAST'. This method with new Expression(..) gives no results. defaultOrder is inactive. Do you have any idea?

**Comment by arogachev** [comment]
Show full code.

**Comment by gadd33** [comment]
```php
        $dataProvider = new ActiveDataProvider([
            'query' => $query,
        	'pagination' => ['pageSize' => 10],
        	'sort'	=> ['defaultOrder' => [new Expression('"finishTime" DESC NULLS LAST')]],
        ]);
```

PHP Notice – yii\base\ErrorException
Undefined offset: 0
1. in C:\wamp\www\zbieracz\vendor\yiisoft\yii2\data\Sort.php at line 213
[...]
213.            $definition = $this->attributes[$attribute];
214.            $columns = $definition[$direction === SORT_ASC ? 'asc' : 'desc'];

**Comment by cebe** [comment]
```php
 $dataProvider = new ActiveDataProvider([
            'query' => $query,
            'pagination' => ['pageSize' => 10],
            'sort'  => [
                'attributes' => [
                    'finishTime' => [
                        'asc' => [new Expression('finishTime NULLS FIRST')],
                        'desc' => [new Expression('finishTime DESC NULLS LAST')],
                   ],
                    // more attribute definitions here
                ],
                'defaultOrder' => ['finishTime' => SORT_DESC],
        ]);
```

**Comment by gadd33** [comment]
Great. Thank you!

**Comment by rakot** [comment]
For MySQL - NULLS LAST
```
$dataProvider->sort->attributes['rank'] = [
     'asc' => [new Expression('`rank` IS NULL ASC, `rank` ASC')],
     'desc' => [new Expression('`rank` IS NULL ASC, `rank` DESC')],
];
```

**Comment by DrDeath72** [comment]
```
 $dataProvider = new ActiveDataProvider([
            'query' => $query,
            'pagination' => ['pageSize' => 10],
            'sort'  => [
                'attributes' => [
                    'finishTime' => [
                        'asc' => [new Expression('finishTime NULLS FIRST')],
                        'desc' => [new Expression('finishTime DESC NULLS LAST')],
                   ],
                    // more attribute definitions here
                ],
                'defaultOrder' => ['finishTime' => SORT_DESC],
        ]);
```
wrong, must be
```
 $dataProvider = new ActiveDataProvider([
            'query' => $query,
            'pagination' => ['pageSize' => 10],
            'sort'  => [
                'attributes' => [
                    'finishTime' => [
                        'asc' => [new Expression('finishTime NULLS FIRST')],
                        'desc' => [new Expression('finishTime DESC NULLS LAST')],
                   ],
                    // more attribute definitions here
                ],
                'defaultOrder' => [new Expression('finishTime DESC NULLS LAST')],
        ]);
```
but it's not work

**Comment by edgardmessias** [comment]
My sugestion:

Make \yii\db\Direction like \yii\db\Expression:

Example:
\yii\db\QueryBuilder:
```php
    public function buildOrderBy($columns)
    {
        if (empty($columns)) {
            return '';
        }
        $orders = [];
        foreach ($columns as $name => $direction) {
            $orders[] = $this->buildOrderByDirection($name, $direction);
        }

        return 'ORDER BY ' . implode(', ', $orders);
    }

    public function buildOrderByDirection($name, $direction)
    {
        if ($direction instanceof Expression) {
            return $direction->expression;
        }
        if ($direction instanceof Direction){
            //non supported DB
            $direction = $direction->direction;
        }
        return $this->db->quoteColumnName($name) . ($direction === SORT_DESC ? ' DESC' : '');
    }
```

For postgresql:
```php
    public function buildOrderByDirection($name, $direction) {
        if ($direction instanceof \yii\db\Direction){
            return $this->db->quoteColumnName($name)
                    . ($direction->direction === SORT_DESC ? ' DESC' : '')
                    . ($direction->nullsFirst ? ' NULLS FIRST' : ' NULLS LAST');
        }
        return parent::buildOrderByDirection($name, $direction);
    }
```

For mysql:
```php
    public function buildOrderByDirection($name, $direction) {
        if ($direction instanceof \yii\db\Direction){
            return  'ISNULL(' . $this->db->quoteColumnName($name) . ')'
                    . ($direction->nullsFirst ? ' DESC' : ' ')
                    . ', ' . $this->db->quoteColumnName($name)
                    . ($direction->direction === SORT_DESC ? ' DESC' : '');
        }
        return parent::buildOrderByDirection($name, $direction);
    }
```

Use:
```php
$query->orderBy([
    'name' => new \yii\db\Direction(SORT_DESC, false),
    'date' => new \yii\db\Direction(SORT_DESC, true)
]);
```
Output:
```sql
-- postgresql
SELECT * FROM "operations" ORDER BY "name" DESC NULLS LAST, "date" DESC NULLS FIRST

--Mysql
SELECT * FROM `operations` ORDER BY ISNULL(`name`), `name` DESC, ISNULL(`date`) DESC, `date` DESC
```

I can make a pull request later

**Comment by klimov-paul** [comment]
Resolved by commit 473598ac8ccbc3f0782e41d0ad2a01abd4b55fb2
